### PR TITLE
@orta: Addresses memory leaks

### LINKED ISSUE #500: Get CI working

### PR DESCRIPTION
**Pull request by ashfurrow**
Tried a fix for #500 but instead just got rid of reference cycles.

### CODE REVIEW

**Comment by ashfurrow** [comment]
Becoming more a general-purpose performance PR. I'm OK with this.

Really glad when we move to a FRP lib that uses the compiler to warn us we forgot `.takeUntil(viewWillDisappearSignal())`.

**Comment by ashfurrow** [comment]
So basically, there are problems with `fulfillmentNavigatonController` being retained everywhere. I've been setting a breakpoint in the nav controller's `deinit` function to make sure it's called under different scenarios. Here's the state of this PR right now:

## Broken

Bid -> Bid Button -> Paddle/Pin -> Bid -> [Cancel]
Bid -> Bid Button -> Paddle/Pin -> Bid -> Place Another bid

## Fixed

Bid -> Cancel
Bid -> Bid Button -> Cancel
Bid -> Bid Button -> Artsy Login
Bid -> Bid Button -> Unknown Phone Number

So now it's just sign-in/bidding that's causing the problem. Will continue.

**Comment by ashfurrow** [comment]
Yesterday, Andy Matuschack said something to the effect of "it's bananas that you can write a reference cycle in Swift" and I think he was right.

**Comment by alloy** [comment]
@ashfurrow It’s kind of inherent with ARC style systems. You _have_ to know what you’re doing, except you don’t need to do all the actual work. You can get a lot better with a garbage collector, but even then it’s still possible. (Did you know the Ruby stdlib comes with a `WeakRef` helper class?)

On whether or not it _should_ be possible I agree, the ideal language is one where you don’t need to.

**Comment by alloy** [comment]
The cycle fixes look good to me, but you should probably let someone that’s used to Swift+RAC take a look as well.

**Comment by ashfurrow** [comment]
Agreed on all counts, thanks for the feedback! Andy's talk was kind of "what might swift have looked like if it didn't need ObjC interop?" A neat thought experiment.

**Comment by ashfurrow** [comment]
The remaining issue is really the following cycle, which affects anything that touches loading (registration, bidding). 

> Fulfillment controller -> Loading VC -> Loading VM -> Fulfillment controller

The fulfillment controller is used primarily to access the logged-in provider. But the provider doesn't exist until the user is logged in. Quite the pickle. 

The tests will need to be updated, too.

**Comment by ashfurrow** [comment]
Phew, ok. 

screen shot 2015-11-02 at 4 32 21 pm

Tests passing locally. I solved the issue mentioned above using `unowned` references, and also spotted a reference cycle in `didSet` for `xAccessToken`, so _any_ time the user was logged in, there was a leak. 

Hopefully Travis will pass. If not, well, there's a Trello card for that, too.

**Comment by ashfurrow** [comment]
Yup, same problem as before. 

@orta this is good to merge now, passes locally. I'll deal with the CI issues separately so we can get this merged.

File: Kiosk/Auction Listings/ListingsCountdownManager.swift [hunk 1]
```diff
@@ -21,6 +23,11 @@ class ListingsCountdownManager: NSObject {
         }
     }
 
+    /// Immediately invalidates all timers. No further updates will be made to the UI after this method is called.
```
Comment (orta) [inline Kiosk/Auction Listings/ListingsCountdownManager.swift]: comment feels like it implies multiple timers, niggle

File: Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift [hunk 2]
```diff
@@ -31,21 +30,20 @@ class ConfirmYourBidArtsyLoginViewController: UIViewController {
...
-
-        let emailTextSignal = emailTextField.rac_textSignal()
-        let passwordTextSignal = passwordTextField.rac_textSignal()
-        RAC(nav.bidDetails.newUser, "email") <~ emailTextSignal.takeUntil(viewWillDisappearSignal())
-        RAC(nav.bidDetails.newUser, "password") <~ passwordTextSignal.takeUntil(viewWillDisappearSignal())
+        
+        let emailTextSignal = emailTextField.rac_textSignal().takeUntil(viewWillDisappearSignal())
+        let passwordTextSignal = passwordTextField.rac_textSignal().takeUntil(viewWillDisappearSignal())
+        RAC(nav.bidDetails.newUser, "email") <~ emailTextSignal
+        RAC(nav.bidDetails.newUser, "password") <~ passwordTextSignal
```
Comment (orta) [inline Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift]: interesting, what's going on here? ( or was this just a style choice? )

File: Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift [hunk 3]
```diff
@@ -31,21 +30,20 @@ class ConfirmYourBidArtsyLoginViewController: UIViewController {
...
         let inputIsEmail = emailTextSignal.map(stringIsEmailAddress)
         let passwordIsLongEnough = passwordTextSignal.map(isZeroLengthString).not()
         let formIsValid = RACSignal.combineLatest([inputIsEmail, passwordIsLongEnough]).and()
 
         confirmCredentialsButton.rac_command = RACCommand(enabled: formIsValid) { [weak self] _ in
-            if (self == nil) {
-                return RACSignal.empty()
-            }
-            return self!.xAuthSignal().`try` { (accessTokenDict, errorPointer) -> Bool in
+            guard let me = self else { return RACSignal.empty() }
```
Comment (orta) [inline Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift]: I like this

File: Kiosk/Bid Fulfillment/ConfirmYourBidViewController.swift [hunk 4]
```diff
@@ -49,36 +48,33 @@ class ConfirmYourBidViewController: UIViewController {
...
+            guard let me = self else { return RACSignal.empty() }
+
+            let endpoint: ArtsyAPI = ArtsyAPI.FindBidderRegistration(auctionID: auctionID, phone: String(me.number))
+            return XAppRequest(endpoint, provider:me.provider).filterStatusCode(400).doError { (error) -> Void in
+                guard let me = self else { return }
+
+                // Due to AlamoFire restrictions we can't stop HTTP redirects
+                // so to figure out if we got 302'd we have to introspect the
+                // error to see if it's the original URL to know if the
+                // request suceedded
```
Comment (orta) [inline Kiosk/Bid Fulfillment/ConfirmYourBidViewController.swift]: typo

File: Kiosk/Bid Fulfillment/FulfillmentNavigationController.swift [hunk 5]
```diff
@@ -24,12 +24,13 @@ class FulfillmentNavigationController: UINavigationController, FulfillmentContro
     /// Otherwise we're fine with a legit auth token
     var xAccessToken: String? {
         didSet(oldToken) {
+            let accessToken = self.xAccessToken!
```
Comment (orta) [inline Kiosk/Bid Fulfillment/FulfillmentNavigationController.swift]: guard?

File: Kiosk/Bid Fulfillment/RegisterViewController.swift [hunk 6]
```diff
@@ -52,17 +51,16 @@ class RegisterViewController: UIViewController {
...
         goToViewController(nextVC)
-
     }
 
     func goToViewController(controller: UIViewController) {
         self.internalNavController()!.viewControllers = [controller]
 
         if let subscribableVC = controller as? RegistrationSubController {
-            subscribableVC.finishedSignal.subscribeCompleted({ [weak self] () -> Void in
+            subscribableVC.finishedSignal.subscribeCompleted { [weak self] () -> Void in
```
Comment (orta) [inline Kiosk/Bid Fulfillment/RegisterViewController.swift]: `Void in` can be removed

File: KioskTests/Bid Fulfillment/PlaceBidNetworkModelTests.swift [hunk 7]
```diff
@@ -1,8 +1,8 @@
 import Quick
 import Nimble
 import ReactiveCocoa
-import Swift_RAC_Macros
-import Moya
+import
```
Comment (orta) [inline KioskTests/Bid Fulfillment/PlaceBidNetworkModelTests.swift]: this works/

**Comment by orta** [comment]
Looks cool to me

File: Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift (hunk 2 above)
Comment (ashfurrow) [inline Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift]: (We talked about this in-person, summarizing here.)

The `emailTextSignal` and `passwordTextSignal` variables are used further down, but the `takeUntil` was only being applied to the first two uses of them, so I moved the call to the variable declaration.

File: Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift (hunk 3 above)
Comment (ashfurrow) [inline Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift]: Have to credit @acjay with this.

File: Kiosk/Bid Fulfillment/RegisterViewController.swift (hunk 6 above)
Comment (ashfurrow) [inline Kiosk/Bid Fulfillment/RegisterViewController.swift]: `in` needs there for `[weak self]`, but taking out `() -> Void` is :+1:

File: KioskTests/Bid Fulfillment/PlaceBidNetworkModelTests.swift (hunk 7 above)
Comment (ashfurrow) [inline KioskTests/Bid Fulfillment/PlaceBidNetworkModelTests.swift]: Ah, yeah. Sorry, project-wide find-and-replace got messy. I was surprised, too.

**Comment by ashfurrow** [comment]
No substantially changes, since CI is not reliable and I got the :+1: I'm going to merge.

File: Kiosk/Bid Fulfillment/BidderNetworkModel.swift [hunk 8]
```diff
@@ -1,12 +1,11 @@
 import Foundation
 import ReactiveCocoa
-import Swift_RAC_Macros
 import Moya
 
 class BidderNetworkModel: NSObject {
     // MARK: - Getters
 
-    var fulfillmentController: FulfillmentController
+    unowned let fulfillmentController: FulfillmentController
```
Comment (acjay) [inline Kiosk/Bid Fulfillment/BidderNetworkModel.swift]: I think even better might be not to inject the entire `FulfillmentController` in the first place, but instead have `FulfillmentController` inject functions that abstract over the queries / side effects the child wants. I've found it to be a really strong pattern is to never have a child component have an explicit reference to a parent. It also makes mocking much simpler -- just pass in a lamba with the assumed behavior.

Comment (ashfurrow) [inline Kiosk/Bid Fulfillment/BidderNetworkModel.swift]: Makes sense – something to keep in mind when I get to https://github.com/artsy/eidolon/issues/433

**Pull request merged**